Sources/PriorityQueueModule/PriorityQueue.swift

```diff
@@ -0,0 +1,363 @@
+//
+//===----------------------------------------------------------------------===//
+
+import Swift
+
+/// A double-ended priority queue built on top of a [Min-Max Heap](https://en.wikipedia.org/wiki/Min-max_heap)
+/// data structure.
+///
+/// In a min-max heap, each node at an even level in the tree is less than all its descendants, while each node
+/// at an odd level in the tree is greater than all of its descendants.
+///
+/// The implementation is based off [this paper](http://akira.ruc.dk/~keld/teaching/algoritmedesign_f03/Artikler/02/Atkinson86.pdf).
+public struct PriorityQueue<Element: Comparable> {
+    private var storage: [Element]
+
+    /// A Boolean value indicating whether or not the queue is empty.
+    ///
+    /// - Complexity: O(1)
+    public var isEmpty: Bool {
+        storage.isEmpty
+    }
+
+    /// The number of elements in the queue.
+    ///
+    /// - Complexity: O(1)
+    public var count: Int {
+        storage.count
+    }
+
+    /// A read-only view into the underlying heap.
+    ///
+    /// In the current implementation, the elements aren't _arbitrarily_ ordered, as a min-max heap
+    /// is used for storage. However, no guarantees are given as to the ordering of the elements.
+    ///
+    /// - Complexity: O(1)
+    public var unordered: [Element] {
+        storage
+    }
+
+    /// Creates an empty queue.
+    public init() {
+        storage = []
+    }
+
+    /// Inserts the given element into the queue.
+    ///
+    /// - Complexity: O(log n)
+    public mutating func insert(_ element: Element) {
+        storage.append(element)
+        _bubbleUp(startingAt: storage.endIndex - 1)
+    }
+
+    /// Returns the element with the lowest priority, if available.
+    ///
+    /// - Complexity: O(1)
+    public func min() -> Element? {
+        storage.first
+    }
+
+    /// Returns the element with the highest priority, if available.
+    ///
+    /// - Complexity: O(1)
+    public func max() -> Element? {
+        switch storage.count {
+        case 0, 1, 2:
+            // If count is 0, `last` will return `nil`
+            // If count is 1, the last (and only) item is the max
+            // If count is 2, the last item is the max (as it's the only item in the first max level)
+            return storage.last
+        default:
+            // We have at least 3 items -- return the larger of the two in the first max level
+            let leftIdx = _leftChildIndex(of: 0)
+            let rightIdx = _rightChildIndex(of: 0)
+
+            let left = storage[leftIdx]
+            let right = storage[rightIdx]
+
+            // Both indexes are valid -- return the larger of the two
+            return Swift.max(left, right)
+        }
+    }
+
+    /// Removes and returns the element with the lowest priority, if available.
+    ///
+    /// - Complexity: O(log n)
+    public mutating func removeMin() -> Element? {
+        return _delete(at: 0)
+    }
```

Comment:
I don't have strong opinions about the names here. @lorentey's [original feedback](https://github.com/apple/swift-collections/pull/43#discussion_r641254919) of @amanuelephrem's PR suggested `removeMin/Max`, so that's why I went with `remove*`. `popMin/Max` seems more in line with the end result of these methods to me, so I'm happy to tweak the naming.

Reply:
`Deque` has both `popFirst`/`popLast` and `removeFirst`/`removeLast` as well, with only the `pop_` methods returning an optional, so it looks to me like the types are now fully consistent with each other in this regard 👍